### DISCUSSION TITLE
Добавить плейлист в общую историю прослушивания

### DISCUSSION THREAD

**Discussion by ilyanikolaich** [Chimildic/goofy#31] [Помощь с алгоритмом] [answered]
У меня есть большой плейлист, в котором хранится история прослушивания треков за несколько месяцев - ровно до момента, как я начал пользоваться Goofy с его историей прослушивания. И сейчас для "вычеркивания" истории я объединяю два списка - историю Goofy и свой плейлист. Хочется упрощения. Хочу добавить _свой_ плейлист истории в общую историю прослушивания Goofy.
Составил такой скрипт, как мне кажется, он должен выполнить мою задачу:

1.получаю свой плейлист в массив треков
2."сжимаю" полученный треклист для экономии места
3.добавляю полученный треклист к существующему записанному на Драйве файлу истории с добавлением в начало.
```javascript
let MyHistoryFull_add = Source.getTracks( [
      { name: 'history_tmp', id: '*******' } 
      ] );
Cache.compressTracks(MyHistoryFull_add);
Cache.append('BothRecentTracks.json', MyHistoryFull_add, 'begin');
```
если я првильно всё понял, то сохранятся даже даты прослушивания треков из моего плейлиста истории - они добавлялись в тот плейлист после прослушивания.

также, учитывая, что история прослушивания Goofy обновляется раз в 15 минут, мне необходимо дождаться очередного завершения выполнения скрипта добавления треков в историю, и после этого запустить совй скрипт, который добавит мой плейлист в общую историю.

имеет ли смысл также добавить сжатие Cache.compressArtists(); после сжатия треков?

Правильно ли я всё понял, и нет ли каких-то неверностей? (не хочу накосячить с историей и сделать всё максимально правильно)
Спасибо.

**Comment by Chimildic** [comment, marked as answer]
Возникает вопрос лимита. Сейчас это 20 тысяч треков. При превышении, наиболее старые вытесняются добавлением новых. Имеет ли смысл этим заниматься, если рано или поздно история затрется новой?

Разметка файла `SpotifyRecentTracks` отличается от разметки массива треков. Грубо говоря, в файле упрощенные ответы Spotify без извлечения треков. Поэтому в этот файл добавить свою историю не получится. Возможно мне стоит привести его к общему знаменателю.

Файл `BothRecentTracks` формируется как совокупность `SpotifyRecentTracks` и `LastfmRecentTracks`. При этом удаляя дубликаты и сортируя по дате. Это происходит **каждый раз**. Файл перезаписывается получившейся совокупностью. Поэтому любая модификация `BothRecentTracks` будет перезаписана при следующем триггере. 

Может быть странно, но для такой задачи подходит файл `LastfmRecentTracks`. Внутри разметка как у массива треков и сами треки формата Spotify. Если в вашем плейлисте истории сортировка от новых к старым (сверху вниз), то их нужно добавить в конец файла `append('LastfmRecentTracks.json', tracks, 'end')`. При этом желательно `compressTracks`. Касаемо `compressArtists` - не нужно. Он для массива исполнителей.

У треков плейлиста дата помечается как `added_at`. У истории прослушиваний `played_at`. Сходу могу ошибиться, но вроде везде они эквиваленты и проблем вызвать не должны. За исключением логики нахождения "новых" прослушиваний в триггере. Но поскольку история плейлиста старая, никак не повлияет.

Также у Google Диска есть история изменений файла. Около 100 правок. Можно откатиться к предыдущему состоянию файла. Или вообще скачать его изначально. И вернуть в случае ошибки.

**Reply by ilyanikolaich** [reply]
Благодарю за подробный развернутый ответ, в том числе и устройству файлов истории вообще. Операция по добавлению треков моего исторического плейлиста в историю ЛастФМ прошла успешно, и теперь общая сводная история состоит из всех общих треков.

**Comment by Chimildic** [comment]
В версии 1.4.0 структура файла `SpotifyRecentTracks` обновлена до обычного массива треков.
Также появилась функция по добавлению треков в историю - [appendTracks](https://chimildic.github.io/goofy/#/func?id=appendtracks).
